File: backend/components/api/wire_types/src/resources/list.rs
use crate::*;
use newtypes::DbActor;
use newtypes::ListAlias;
use newtypes::ListEntryId;
use newtypes::ListId;
use newtypes::ListKind;
use newtypes::ObConfigurationId;
use newtypes::PiiString;

#[derive(Debug, Clone, Serialize, Apiv2Response, macros::JsonResponder)]
pub struct List {
    pub id: ListId,
    pub name: String,
    pub alias: ListAlias,
    pub kind: ListKind,
    pub created_at: DateTime<Utc>,
    pub actor: DbActor,
    pub entries_count: usize,
    pub used_in_playbook: bool,
}

#[derive(Debug, Clone, Serialize, Apiv2Response, macros::JsonResponder)]
pub struct ListDetails {
    pub id: ListId,
    pub name: String,
    pub alias: ListAlias,
    pub kind: ListKind,
    pub created_at: DateTime<Utc>,
    pub actor: DbActor,
    pub playbooks: Vec<ListPlaybookUsage>,
}
#[derive(Debug, Clone, Serialize, Apiv2Schema)]
pub struct ListEntry {
    pub id: ListEntryId,
    pub data: PiiString,
    pub created_at: DateTime<Utc>,
    pub actor: DbActor,
}

#[derive(Debug, Clone, Serialize, Apiv2Schema)]
pub struct ListPlaybookUsage {
    pub id: ObConfigurationId,
    pub name: String,
    pub rules: Vec<Rule>,
}

File: frontend/packages/types/src/data/list.ts
import type { Actor } from './actor';
import type { InsightEvent } from './insight-event';
import type { Rule } from './rule';

export enum ListKind {
  emailAddress = 'email_address',
  emailDomain = 'email_domain',
  ssn9 = 'ssn9',
  phoneNumber = 'phone_number',
  phoneCountryCode = 'phone_country_code',
  ipAddress = 'ip_address',
}

export type ListPlaybookUsage = {
  id: string;
  key: string;
  name: string;
  rules: Rule[];
};

export type List = {
  id: string;
  actor: Actor;
  alias: string;
  entriesCount: number;
  usedInPlaybook: boolean;
  createdAt: string;
  kind: ListKind;
  name: string;
};

export type ListDetails = {
  id: string;
  actor: Actor;
  alias: string;
  createdAt: string;
  kind: ListKind;
  name: string;
  playbooks: ListPlaybookUsage[];
};

export type ListEntry = {
  id: string;
  data: string;
  createdAt: string;
  actor: Actor;
};

export enum ListTimelineEventKind {
  createList = 'create_list',
  updateList = 'update_list',
  createListEntry = 'create_list_entry',
  deleteListEntry = 'delete_list_entry',
}

// TODO: implement after backend is ready
export type ListCreatedEvent = {
  kind: ListTimelineEventKind.createList;
  data: unknown;
};

// TODO: implement after backend is ready
export type ListUpdatedEvent = {
  kind: ListTimelineEventKind.updateList;
  data: unknown;
};

export type ListEntryCreatedEvent = {
  kind: ListTimelineEventKind.createListEntry;
  data: {
    listEntryCreationId: string;
    entries: string[];
  };
};

export type ListEntryDeletedEvent = {
  kind: ListTimelineEventKind.deleteListEntry;
  data: {
    listEntryId: string;
    entry: string;
  };
};

export type ListTimelineEvent = {
  id: string;
  timestamp: string;
  tenantId: string;
  name: ListTimelineEventKind;
  principal: {
    kind: string;
    member: string;
  };
  insightEvent: InsightEvent;
  detail: ListCreatedEvent | ListUpdatedEvent | ListEntryCreatedEvent | ListEntryDeletedEvent;
};

export type ListTimeline = ListTimelineEvent[];
